bitcoin/src/psbt/serialize.rs

```diff
@@ -384,23 +387,22 @@ impl Deserialize for (Vec<TapLeafHash>, KeySource) {
 
 impl Serialize for TapTree {
     fn serialize(&self) -> Vec<u8> {
-        match (self.0.branch().len(), self.0.branch().last()) {
-            (1, Some(Some(root))) => {
-                let mut buf = Vec::new();
-                for leaf_info in root.leaves.iter() {
-                    // # Cast Safety:
-                    //
-                    // TaprootMerkleBranch can only have len atmost 128(TAPROOT_CONTROL_MAX_NODE_COUNT).
-                    // safe to cast from usize to u8
-                    buf.push(leaf_info.merkle_branch().as_inner().len() as u8);
-                    buf.push(leaf_info.leaf_version().to_consensus());
-                    leaf_info.script().consensus_encode(&mut buf).expect("Vecs dont err");
-                }
-                buf
-            }
-        // This should be unreachable as we Taptree is already finalized
-            _ => unreachable!(),
+        let capacity = self.script_leaves().map(|l|
+                l.merkle_branch().len()* 32 + // merkle branch
+                l.script().len() + VarInt(l.script().len() as u64).len() + // script version
+                1 // leaf version
+            ).sum::<usize>();
+        let mut buf = Vec::with_capacity(capacity);
+        for leaf_info in self.script_leaves() {
+            // # Cast Safety:
+            //
+            // TaprootMerkleBranch can only have len atmost 128(TAPROOT_CONTROL_MAX_NODE_COUNT).
+            // safe to cast from usize to u8
+            buf.push(leaf_info.merkle_branch().len() as u8);
```

Comment:
Why the `* 32` when we push just the length here?

Reply:
another good catch. I really need to think before I am typing :(